### PR TITLE
Apply overrides for data_input_dict in Node

### PR DESCRIPTION
**Pull request by liuly12**
Please note these changes do no apply to `Surface` and its superclasses as they are based on the `DecayTank`, which does not have `data_input_dict` attribute. The overrides for `data_input_dict` in `Surface` are addressed in issue #76.

### CODE REVIEW

File: tests/test_nodes.py
Comment (barneydobson) [inline tests/test_nodes.py:678]: I think would be tidier with `tempfile` (and you can also delete the example data dict file):

```suggestion
        from pathlib import Path
        import tempfile
        with tempfile.TemporaryDirectory as temp_dir:
                fid = Path(temp_dir) / "example_data_input_dict.csv"
                pd.DataFrame(new_data_input_dict).to_csv(fid)
                node.apply_overrides({'data_input_dict': fid})
                self.assertDictEqual(node.data_input_dict, new_data_input_dict)
```

File: wsimod/nodes/nodes.py
Comment (barneydobson) [inline wsimod/nodes/nodes.py:98]: Maybe I misunderstand but don't we need something in case they are not overriding `data_input_dict` or that node doesn't use `data_input_dict` ?

```suggestion
        elif not content:
                pass
        else:
```

**Review by barneydobson — APPROVED** [review]
Couple points here - if you're happy with my suggestions please just double check the tests still run and there's no typos before merging

File: tests/test_nodes.py
Comment (liuly12) [reply tests/test_nodes.py:678]: an error raises
```
    with tempfile.TemporaryDirectory as temp_dir:
AttributeError: __enter__
```

**Review by barneydobson — APPROVED** [review]

File: tests/test_nodes.py
Comment (barneydobson) [inline tests/test_nodes.py:674]:
```suggestion
        new_data_input_dict = str(Path(__file__).parent / "example_data_input_dict.csv.gz")
```

File: tests/test_nodes.py
Comment (barneydobson) [inline tests/test_nodes.py:24]:
```suggestion
from pathlib import Path
```

**Review by barneydobson — CHANGES_REQUESTED** [review]
fixed the load error I think

File: tests/test_nodes.py
Comment (liuly12) [reply tests/test_nodes.py]: Tried but still not passed ...

Comment (barneydobson) [reply tests/test_nodes.py]: It looks like that's just the pre commit since my code suggestion doesn't fit the formatting rules. If you pull, run `pre-commit run --all-files`, add, commit, push, then hopefully it will pass

Comment (liuly12) [reply tests/test_nodes.py]: passed for the qa but not for windows-latest 3.10, here is the error message:
'Run python -m pytest
C:\hostedtoolcache\windows\Python\3.[10](https://github.com/ImperialCollegeLondon/wsi/actions/runs/10190065866/job/28189182286#step:5:11).11\x64\python.exe: No module named pytest
Error: Process completed with exit code 1.'

Comment (barneydobson) [reply tests/test_nodes.py]: what in the world... super weird... I'm happy to approve - not sure if @dalonsoa may have views though - will see when he's back

**Review by barneydobson — APPROVED** [review]

File: tests/test_nodes.py
Comment (dalonsoa) [reply tests/test_nodes.py]: For Python 3.9 and 3.11, the dependencies are all installed fine and you get at the end of the `Install and build` step in the CI file something like:
```
Installing collected packages: pytz, distlib, wheel, tzdata, six, ruff, pyyaml, pyproject-hooks, pyparsing, pluggy, platformdirs, pillow, pathspec, packaging, numpy, nodeenv, mypy-extensions, kiwisolver, iniconfig, identify, fonttools, filelock, dill, cycler, coverage, colorama, cfgv, certifi, attrs, virtualenv, tqdm, shapely, python-dateutil, pytest, pyproj, contourpy, click, build, pytest-cov, pre-commit, pip-tools, pandas, matplotlib, cligj, click-plugins, black, fiona, geopandas
Successfully installed attrs-23.2.0 black-23.10.1 build-1.0.3 certifi-2023.11.17 cfgv-3.4.0 click-8.1.7 click-plugins-1.1.1 cligj-0.7.2 colorama-0.4.6 contourpy-1.2.0 coverage-7.3.2 cycler-0.12.1 dill-0.3.7 distlib-0.3.7 filelock-3.13.1 fiona-1.9.5 fonttools-4.47.0 geopandas-0.14.2 identify-2.5.31 iniconfig-2.0.0 kiwisolver-1.4.5 matplotlib-3.8.2 mypy-extensions-1.0.0 nodeenv-1.8.0 numpy-1.26.2 packaging-23.2 pandas-2.1.3 pathspec-0.11.2 pillow-10.2.0 pip-tools-7.3.0 platformdirs-3.11.0 pluggy-1.3.0 pre-commit-3.5.0 pyparsing-3.1.1 pyproj-3.6.1 pyproject-hooks-1.0.0 pytest-7.4.3 pytest-cov-4.1.0 python-dateutil-2.8.2 pytz-2023.3.post1 pyyaml-6.0.1 ruff-0.1.4 shapely-2.0.2 six-1.16.0 tqdm-4.66.1 tzdata-2023.3 virtualenv-20.24.6 wheel-0.41.3
```

But for Python 3.10 with Windows in the `pull_request` scenario you get:

```
...
Downloading colorama-0.4.6-py2.py3-none-any.whl (25 kB)
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
    unknown package:
        Expected sha256 5b4c6a3df53bee8f85bb46685562b21b43346be1fe96419f18f70fa1ab8c561c
             Got        07ce221ab790f1ee6963ea46b7aef0e53e18dbc9c77d48075e7e4363dd02213c
...
```

And then, a bit further down, 

```
Installing collected packages: pytz, tzdata, six, PyYAML, numpy, dill, colorama, tqdm, python-dateutil, pandas, wsimod
Successfully installed PyYAML-6.0.1 colorama-0.4.6 dill-0.3.8 numpy-2.0.1 pandas-2.2.2 python-dateutil-2.9.0.post0 pytz-2024.1 six-1.16.0 tqdm-4.66.4 tzdata-2024.1 wsimod-0.4.0
```

So, indeed, `pytest` is not installed - and neither many other packages. It has nothing to do with using `pathlib` or with `pre-commit` but with that hash that is not what it should. 

I think it must have been a glitch or something because I've re-run the failed job and now it passes fine.

When this kind of very weird things happen, just re-run the jobs a couple of times - if some time in between - because GitHub actions, like any other software, do have errors and temporary issues that fix themselves if given some time.

File: wsimod/nodes/nodes.py
Comment (dalonsoa) [inline wsimod/nodes/nodes.py:98]: Why these two cases? Wouldn't it be best to tell the user to put the data input dict directly in the overrides? Or only in an external file. 

I might be a bit cynic here, but I think that the more freedom you give a user with the configuration, the higher the chance for them to do it wrong.

File: tests/test_nodes.py
Comment (dalonsoa) [inline tests/test_nodes.py:682]: No print statements in tests. If you want to see what's going on, use a debugger. 
```suggestion
```

File: tests/test_nodes.py
Comment (dalonsoa) [inline tests/test_nodes.py:678]: You need some consistency with the imports. Either import everything at the top of the file or at the top of the function, but not in the middle of it - unless there is a very good reason, like using patching and things like that, which is not the case.

**Review by dalonsoa — CHANGES_REQUESTED** [review]
A couple of areas of improvement.

File: wsimod/nodes/nodes.py
Comment (liuly12) [reply wsimod/nodes/nodes.py:97]: The reason behind is that I don't know what preferences the users might have. I personally think directly using read_csv would be a more reasonable way - users can specify the timeseries csv files and only provide the address here. Our colleagues often have problems in directly changing the data_input_dict because the timestamp in the key of the dict (e.g., `(precipitation, 2000-01-01)`) is a new class defined in WSIMOD, so that they often get the error: `key *** is not found` when they specify the timstamp as a normal datetime. I don't know whether @barneydobson has a preference here.

File: tests/test_nodes.py
Comment (liuly12) [reply tests/test_nodes.py]: I'll wait the conversation on `read_csv` to be resolved to address this.

Comment (liuly12) [reply tests/test_nodes.py]: @dalonsoa If I move this `from wsimod.orchestration.model import read_csv` outsidethe `apply_overrides` function to the beginning, I will have the following error which breaks 11 `test` scripts.

```
Traceback (most recent call last):

  File "C:\Users\ll6715\Anaconda3\envs\wsimod\lib\site-packages\spyder_kernels\py3compat.py", line 356, in compat_exec
    exec(code, globals, locals)

  File "c:\users\ll6715\documents\github\wsi\tests\test_nodes.py", line 14, in <module>
    from wsimod.nodes.nodes import Node

  File "c:\users\ll6715\documents\github\wsi\wsimod\nodes\__init__.py", line 2, in <module>
    from wsimod.nodes.catchment import Catchment

  File "c:\users\ll6715\documents\github\wsi\wsimod\nodes\catchment.py", line 9, in <module>
    from wsimod.nodes.nodes import Node

  File "c:\users\ll6715\documents\github\wsi\wsimod\nodes\nodes.py", line 13, in <module>
    from wsimod.orchestration.model import read_csv

  File "c:\users\ll6715\documents\github\wsi\wsimod\orchestration\model.py", line 21, in <module>
    from wsimod.nodes.land import ImperviousSurface

  File "c:\users\ll6715\documents\github\wsi\wsimod\nodes\land.py", line 12, in <module>
    from wsimod.nodes.nodes import Node

ImportError: cannot import name 'Node' from partially initialized module 'wsimod.nodes.nodes' (most likely due to a circular import) (c:\users\ll6715\documents\github\wsi\wsimod\nodes\nodes.py)
```

But if it stays in the function, no errors reported from the tests.

Comment (dalonsoa) [reply tests/test_nodes.py]: So don't move it. In this particular case, it needs to be where it is, otherwise, there are circular imports. The general rule is "at the beginning", but there are reasons for exceptions.

File: wsimod/nodes/nodes.py
Comment (dalonsoa) [inline wsimod/nodes/nodes.py:99]: This should be an exception, so the program halts.
```suggestion
            raise RuntimeError("Not recognised format for data_input_dict")
```
If, for any reason, you do not want to consider it a catastrophic failure and the program can keep going, then you should use a logger with either `error` or `warning`:

```python 
from logging import getLogger

getLogger().error("Not recognised format for data_input_dict")
```

File: tests/test_nodes.py
Comment (dalonsoa) [inline tests/test_nodes.py:422]: This only test the first branch of the conditional statement. You need another test for the case of `not content` and another one for the `else` case, which is an error.

**Review by dalonsoa — CHANGES_REQUESTED** [review]
This looks ok, but it needs a couple of tweaks and extra tests.

File: tests/test_nodes.py
Comment (liuly12) [reply tests/test_nodes.py:422]: Will add one. Just wonder how to verify the error message in the test function?

File: tests/test_nodes.py
Comment (barneydobson) [reply tests/test_nodes.py]: oops needed `tempfile.TemporaryDirectory()` - but I'm happy to ignore it as is only a test

**Review by barneydobson — APPROVED** [review]

**Pull request merged**
